sdk/daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala

```diff
@@ -1139,41 +1141,198 @@ private[lf] object SBuiltin {
+                      case Some(dstArg) =>
+                        viewInterface(machine, interfaceId, dstTplId, dstArg) { dstView =>
+                          executeExpression(machine, SEPreventCatch(srcView)) { srcViewValue =>
+                            // If the destination and src templates are the same, we skip the computation
+                            // of the destination template's view.
+                            if (dstTplId == srcTplId)
+                              cacheContractAndReturnAny(machine, coid, dstTplId, dstArg)(k)
+                            else
+                              executeExpression(machine, SEPreventCatch(dstView)) { dstViewValue =>
+                                if (srcViewValue != dstViewValue) {
+                                  Control.Error(
+                                    IE.Dev(
+                                      NameOf.qualifiedNameOfCurrentFunc,
+                                      IE.Dev.Upgrade(
+                                        IE.Dev.Upgrade.ViewMismatch(
+                                          coid,
+                                          interfaceId,
+                                          srcTplId,
+                                          dstTplId,
+                                          srcView = srcViewValue.toUnnormalizedValue,
+                                          dstView = dstViewValue.toUnnormalizedValue,
+                                        )
+                                      ),
+                                    )
+                                  )
+                                } else
+                                  cacheContractAndReturnAny(machine, coid, dstTplId, dstArg)(k)
+                              }
+                          }
+                        }
+                    }
+                  }
+                }
+              }
+            }
+          }
+      }
+    }
+  }
+
+  private[this] def resolvePackageName[Q](machine: UpdateMachine, pkgName: Ref.PackageName)(
+      k: PackageId => Control[Q]
+  ): Control[Q] = {
+    machine.packageResolution.get(pkgName) match {
+      // TODO https://github.com/digital-asset/daml/issues/17995
+      //  We need a proper interpretation error here
+      case None => crash(s"cannot resolve package $pkgName")
+      case Some(pkgId) => k(pkgId)
+    }
+  }
+
+  // Precondition: the package of tplId is loaded in the machine
+  private[this] def ensureTemplateImplementsInterface[Q](
+      machine: Machine[_],
+      tplId: TypeConName,
+      ifaceId: TypeConName,
+  )(k: () => Control[Q]): Control[Q] = {
+    val tplImplementsInterface =
+      machine.compiledPackages.pkgInterface
+        .lookupTemplate(tplId)
+        .exists(_.implements.contains(ifaceId))
+    if (tplImplementsInterface) k()
+    else crash(s"template $tplId does not implement interface $ifaceId")
```

Comment:
Should this be a crash ?

Reply:
if yes, please add a comment.

Reply:
Turns out `castAnyInterface` already does that check and its result is never used. So I:
- deleted this flawed version of `ensureTemplateImplementsInterface`
- slightly modified `castAnyInterface` to return no result 
- renamed  `castAnyInterface` to `ensureTemplateImplementsInterface`
- deleted the `SBCastAnyInterface` builtin that was no longer used